### ISSUE TITLE
Do not hard error if libfuse is not present

### ISSUE THREAD

**Issue by probonopd** [AppImage/AppImageKit#123]
We could use `dlopen ("/usr/lib64/libfuse.so.2", RTLD_LAZY);` like on http://www.dwheeler.com/program-library/Program-Library-HOWTO/x172.html to print a nice message saying you can extract or mount the ISO if libfuse is not present.

Message:
It looks like libfuse is not installed on your system. To run this AppImage, you can loop-mount it or extract its contents, e.g. 7zx. See https://github.com/probonopd/AppImageKit/wiki/FUSE for more information.

**Comment by RazZziel** [comment]
Very interesting. According to the manual, `dlopen("libfuse.so.2", RTLD_LAZY);` would be more advisable, because it'd look for the library in the standard paths, although a fallback could be setup to find the library for instance in /tmp/libfuse.so.2, where we can download it without messing up the system, specially on 32bit AppImages on 64bit systems, where 32bit libfuse.so.2 is difficult to install.

**Comment by hideout** [comment]
nice dlopen sounds good for that problem

2016-03-30 7:30 GMT+02:00 Ismael Barros² <notifications@github.com>:

**Comment by probonopd** [comment]
> dlopen("libfuse.so.2", RTLD_LAZY); would be more advisable

Correct. Anyone volunteers to do a PR?

**Comment by crayxt** [comment]
Will this cover the case when user is not a member of fuse group?

**Comment by RazZziel** [comment]
I don't think so, but at least we can warn the user

**Comment by probonopd** [comment]
KDevelop IRC thinks this would be a good idea, since a good error message helps in these cases.

**Comment by probonopd** [comment]
Here is an example of how to do this:
https://github.com/Efreak/apt-cacher-ng/blob/bb751da674c03ea81f0b4381c606b537c853b679/fs/httpfs.cc#L783-L807

Now, how do I get this into https://github.com/probonopd/appimagetool/blob/master/runtime.c? Probably it has to go into squashfuse... [hl.c](https://github.com/vasi/squashfuse/blob/78360b7af4911d46593ce50874ed2fc5f2a6848b/hl.c) looks similar to the one linked above in that it calls `fuse_main()` but we are using [ll.c](https://github.com/vasi/squashfuse/blob/78360b7af4911d46593ce50874ed2fc5f2a6848b/ll.c) iirc - so probably we need to `dlsym` significantly more than in the example above.

**Comment by darealshinji** [comment]
I have an idea:
You could build the main part of the runtime as a shared library (everything requiring libfuse). Create a header file from it (`xxd -i AppImage.so > AppImage_so.h`), include it in the the now smaller runtime binary, and when running the binary it will dump the library into a temporary file and try to dlopen() it. If it fails because libfuse is missing, print something useful.

**Comment by probonopd** [comment]
Cool idea but makes the code larger and more complex, and is not so clean (clutters `/tmp`).

What do you think, could we somehow cleverly use [weak symbols](https://en.wikipedia.org/wiki/Weak_symbol)?

**Comment by darealshinji** [comment]
I gave it a try: https://github.com/darealshinji/AppImageKit/tree/dlopen
The binaries are linked against libdl and not libfuse, but I didn't check if the runtime actually works when you build an AppImage with it.

**Comment by probonopd** [comment]
Can you upload a build here? Thanks.

**Comment by darealshinji** [comment]
[build.zip](https://github.com/probonopd/AppImageKit/files/800206/build.zip)

**Comment by probonopd** [comment]
@darealshinji hats off, it works :+1: :100: :1st_place_medal: 

Please send PR.

**Comment by darealshinji** [comment]
Whoa, nice. Can you check what happens if libfuse is not present? Right now the invoked macros will call `exit(1)`. I don't know if that's ok. Also, what kind of error message do you want? Something including a link to the [wiki](https://github.com/probonopd/AppImageKit/wiki/FUSE) maybe?

**Comment by probonopd** [comment]
While you are at it, this is the error message that appears when the library is not available:

```
me@host:~$ ./XChat_IRC-x86_64.AppImage 
dlopen(): error loading libfuse.so.2
execv error: No such file or directory
```

But even in this case, this __works__:
```
me@host:~$ ./XChat_IRC-x86_64.AppImage --appimage-extract
```

As the message, how about [this](https://github.com/probonopd/AppImageKit/blob/b5ece7a5d249185137034b7162c90917c1e3ef76/runtime.c#L388-L392)?

**Comment by probonopd** [comment]
When libfuse is not present, we currently get

```
me@host:~$ /home/me/Downloads/appimagetool-x86_64.AppImage 
dlopen(): error loading libfuse.so.2
AppImages require FUSE to run. 
You might still be able to extract the contents of this AppImage 
if you run it with the --appimage-extract option. 
See https://github.com/probonopd/AppImageKit/wiki/FUSE 
for more information

execv error: No such file or directory
```

Can you get rid of the `execv error: No such file or directory`?

**Comment by darealshinji** [comment]
https://github.com/probonopd/AppImageKit/pull/365

**Comment by TheAssassin** [comment]
This has been solved in #365.
